html/captive-portal/lib/captiveportal/PacketFence/Controller/TLSProfile.pm

```diff
@@ -0,0 +1,242 @@
+        $self->showError($c,"An error has occured while trying to save your certificate, please contact your IT support");
+    }
+    print $fh "$cert_data\n";
+    close $fh;
+}
+
+sub get_cert : Private {
+    my ($self, $c) = @_;
+    my $portalSession = $c->portalSession;
+    my $stash = $c->stash;
+    my $mac           = $portalSession->clientMac;
+    my $provisioner   = $c->profile->findProvisioner($mac);
+    return unless $provisioner;
+    my $pki_provider = $provisioner->getPkiProvider();
+    my $cert_content = $pki_provider->get_cert({ certificate_email => $stash->{certificate_email}, certificate_cn => $stash->{certificate_cn}, certificate_pwd => $stash->{certificate_pwd} });
+    $c->log->debug("cert_content from pki service $cert_content");
+    $c->stash(cert_content => $cert_content);
+}
+
+sub cert_process : Local {
+    my ($self,$c) = @_;
+    my $logger = $c->log;
+    $c->stash(info => $c->session->{info});
+    $c->forward('validateform');
+    $c->forward('get_cert');
+    $c->forward('build_cert_p12');
+    $c->forward('b64_cert');
+    $c->forward('export_fingerprint');
+    $c->forward( 'Authenticate' => 'checkIfProvisionIsNeeded' );
+    $c->forward( 'CaptivePortal' => 'webNodeRegister', [$c->stash->{info}{pid}, %{$c->stash->{info}}]);
+    $c->forward( 'CaptivePortal' => 'endPortalSession' );
+}
+
+sub validateform : Private {
+    my ($self, $c) = @_;
+    my $logger = $c->log;
+    my $pid    = "";
+    my $portalSession = $c->portalSession;
+    my $mac    = $portalSession->clientMac;
+    unless ($c->has_errors) {
+        my $mac           = $portalSession->clientMac;
+        my $node_info     = node_view($mac);
+        my $pid           = $node_info->{'pid'};
+    }
+    my $passwd1 = $c->request->param('certificate_pwd');
+    my $passwd2 = $c->request->param('certificate_pwd_check');
+    if($passwd1 ne $passwd2) {
+        $c->stash(txt_validation_error => 'Passwords do not match');
+        $c->detach('index');
+    }
+    my $certificate_cn = $mac;
+    $certificate_cn =~ s/:/-/g;
+    my $user_cache = $c->user_cache;
+    my $pki_session = {
+        service           => $c->request->param('service'),
+        certificate_cn    => $certificate_cn,
+        certificate_email => $c->request->param('certificate_email'),
+        certificate_pwd   => $c->request->param('certificate_pwd'),
+    };
+    $user_cache->set("pki_session" => $pki_session);
+    $c->stash($pki_session);
+
+    #unless (Email::Valid->address($email_addr)){
+    #    $logger->debug("Email enter is invalid for username \"$pid\"");
+    #    $self->showError($c,"Please enter a vaild email address");
+    #}
+}
+
+sub b64_cert : Local {
+    my ($self,$c) = @_;
+    my $logger = $c->log;
+    my $session = $c->session;
+    my $stash = $c->stash;
+    my $cwd = $cert_dir;
+    my $certfile = $stash->{'certificate_cn'};
+    my $certp12 = "$cwd/$certfile.p12";
+    my $b64 = pf_run("base64 $certp12");
```

Comment:
You should open that file in perl and convert it to base64 using a perl library not calling the OS for that